debugger/show_source.ml

```diff
@@ -23,18 +23,23 @@ open Source
 
 (* Print a line; return the beginning of the next line *)
 let print_line buffer line_number start point before =
-  let next = next_linefeed buffer start
+  let next_linefeed = next_linefeed buffer start
   and content = buffer_content buffer
   in
     printf "%i " line_number;
+    let next =
+      if content.[next_linefeed - 1] = '\r' then
+        next_linefeed - 1
+      else
+        next_linefeed in
     if point <= next && point >= start then
       (print_string (String.sub content start (point - start));
        print_string (if before then event_mark_before else event_mark_after);
        print_string (String.sub content point (next - point)))
     else
       print_string (String.sub content start (next - start));
     print_newline ();
-    next
+    next_linefeed
```

Comment:
I found this whole code confusing (defining `next` as `next_linefeed - 1` looks like you want to *keep* `\r` instead of get rid of it to someone that has not read the code in depth.), plus I'm not fond of shadowing a function name with the name of its return value. I would suggest the following: define `let linefeed = next_linefeed buffer start`, and rename your current `next` variable into `upto_linefeed` or `before_linefeed` or `line_limit` or `line_end` -- pre-announcing its use as a cutoff point for a `sub` call.